src/fsharp/CheckExpressions.fs

```diff
@@ -4392,6 +4396,59 @@ and TcTypeOrMeasure optKind cenv newOk checkCxs occ env (tpenv: UnscopedTyparEnv
     | SynType.Paren(innerType, _) ->
         TcTypeOrMeasure optKind cenv newOk checkCxs occ env (tpenv: UnscopedTyparEnv) innerType
 
+and TcErasedUnionTypeOr (cenv: cenv) env (tpenv: UnscopedTyparEnv) synCases m =
+    let g = cenv.g
+    // Helper method for eliminating duplicate types from lists of types that form a union type,
+    // create a disjoint set of cases
+    // taking into account that a subtype is a "duplicate" of its supertype.
+    let rec addToCases (pt: TType) (list: ResizeArray<TType>) =
+        if not (ResizeArray.exists (isObjTy g) list) then
+            if isObjTy g pt then
+                list.Clear()
+                list.Add(pt)
```

Comment:
This creates a compatibility consideration for libraries that publish erased union types. It might not be a concern, but I think it's important to be aware of:

Library1
```fsharp
type Foo = obj
```

Library2 (references Library1)
```fsharp
type Bar = (* ... *)
type Problematic = (Foo | Bar) // where `Foo` is from Library1
```

Due to this conditional, the `Bar` part of `Problematic` will be discarded, keeping only `obj` aliased through `Foo`. But if Library1 is later updated to change the type of `Foo`, any code compiled against the updated Library1 and old Library2 will find that the `Problematic` type no longer accepts `Bar`s.

Reply:
Wouldn't that be same issue as the following:

Library1
```fsharp
type Foo = int
```
Library2 (references Library1)

```fsharp
let handle (foo: Foo) = 
   foo + 1
```

Later Library1 is updated to so that `Foo` is aliased to `String`?

Reply:
Yeah you'd definitely expect that sort of breakage with the `Foo` type. But I wouldn't expect that using `Problematic` with `Bar` would also be broken in that case.

Reply:
But you would just recompile the code no in that case?

Reply:
Sure, if you could. But it might be a dependency you don't have the source for

Reply:
I don't see, how this is different to other issues that arise out of dependency version mismatch. You'd only have the issue if you publish a package using a version and then republish it using the same version, causing the breakage.